File: cli/internal/sanitizer/sanitizer.go
// ── CONDUIT MANAGED FILE ────────────────────────────────────────────
// file:        cli/internal/sanitizer/sanitizer.go
// description: CLI-side fail-closed wrapper around the TypeScript sanitizer subprocess.
// owner:       BOTH
// update:      Manual when ingress sanitization behavior changes.
// schema:      security/sanitizer/patterns.yaml
// last_update: 2026-04-04
// ─────────────────────────────────────────────────────────────────────
package sanitizer

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

type Result struct {
	Allowed   bool
	Sanitized string
	Decision  string
	Patterns  []string
	LogEntry  string
}

type cliResponse struct {
	Allowed   bool     `json:"allowed"`
	Sanitized string   `json:"sanitized"`
	Decision  string   `json:"decision"`
	Matches   []string `json:"matches"`
}

func Enabled() bool {
	return true
}

func Sanitize(commandName, input string) (*Result, error) {
	repoRoot, err := findRepoRoot()
	if err != nil {
		return nil, failClosed(repoRoot, commandName, input, "sanitizer_unavailable")
	}

	scriptPath := filepath.Join(repoRoot, "dist", "security", "sanitizer", "cli.js")
	if _, err := os.Stat(scriptPath); err != nil {
		return nil, failClosed(repoRoot, commandName, input, "sanitizer_unavailable")
	}

	cmd := exec.Command("node", scriptPath, commandName)
	cmd.Dir = repoRoot
	cmd.Stdin = strings.NewReader(input)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return nil, failClosed(repoRoot, commandName, input, "sanitizer_unavailable")
	}

	var response cliResponse
	if err := json.Unmarshal(output, &response); err != nil {
		return nil, failClosed(repoRoot, commandName, input, "sanitizer_parse_error")
	}

	decision := mapDecision(response.Decision)
	result := &Result{
		Allowed:   response.Allowed,
		Sanitized: response.Sanitized,
		Decision:  decision,
		Patterns:  response.Matches,
	}

	logEntry, logErr := appendLog(repoRoot, commandName, input, decision, response.Matches)
	if logErr == nil {
		result.LogEntry = logEntry
	}

	if !result.Allowed {
		return result, fmt.Errorf("CONDUIT: sanitizer blocked content for %s", commandName)
	}

	return result, nil
}

func failClosed(repoRoot, commandName, input, pattern string) error {
	if repoRoot != "" {
		_, _ = appendLog(repoRoot, commandName, input, "block", []string{pattern})
	}
	return fmt.Errorf("sanitizer unavailable: fail closed")
}

func mapDecision(decision string) string {
	switch decision {
	case "block_and_escalate":
		return "block"
	case "sanitize_and_log":
		return "sanitize"
	default:
		return "allow"
	}
}

func findRepoRoot() (string, error) {
	wd, err := os.Getwd()
	if err != nil {
		return "", err
	}

	current := wd
	for {
		if _, statErr := os.Stat(filepath.Join(current, ".conduit")); statErr == nil {
			return current, nil
		}
		parent := filepath.Dir(current)
		if parent == current {
			return "", fmt.Errorf("repo root not found")
		}
		current = parent
	}
}

func appendLog(repoRoot, commandName, input, decision string, patterns []string) (string, error) {
	hash := sha256.Sum256([]byte(input))
	entry := fmt.Sprintf(
		"%s command=%s input_sha256=%s decision=%s patterns=%s",
		time.Now().UTC().Format(time.RFC3339),
		commandName,
		hex.EncodeToString(hash[:]),
		decision,
		strings.Join(patterns, ","),
	)

	logPath := filepath.Join(repoRoot, ".conduit", "sanitizer.log")
	file, err := os.OpenFile(logPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return "", err
	}
	defer file.Close()

	if _, err := fmt.Fprintln(file, entry); err != nil {
		return "", err
	}

	return entry, nil
}

File: cli/src/internal/sanitizer.ts
// ── CONDUIT MANAGED FILE ────────────────────────────────────────────
// file:        cli/src/internal/sanitizer.ts
// description: Ingress sanitizer. Reads patterns.yaml and applies them inline.
// owner:       BOTH
// update:      Manual when ingress sanitization behavior changes.
// schema:      security/sanitizer/patterns.yaml
// last_update: 2026-04-07
// ─────────────────────────────────────────────────────────────────────

import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import yaml from 'js-yaml';

interface PatternGroup {
  description: string;
  severity: 'HIGH' | 'MEDIUM' | 'LOW';
  patterns?: string[];
  max_input_chars?: number;
  action?: string;
}

interface PatternsFile {
  version: string;
  patterns: Record<string, PatternGroup>;
  actions: Record<string, string>;
}

export interface SanitizeResult {
  allowed: boolean;
  sanitized: string;
  decision: string;
  matches: string[];
}

/** Locate the conduit install root (where this CLI lives), independent of
 *  the caller's cwd. patterns.yaml lives here, never in the target repo —
 *  running `conduit convoy new` from any target repo must still reach
 *  conduit's sanitizer rules. dist/cli/src/internal/sanitizer.js → four dirs up. */
function getConduitInstallRoot(): string {
  return path.resolve(__dirname, '..', '..', '..', '..');
}

export function sanitize(commandName: string, input: string, repoRoot?: string): SanitizeResult {
  // repoRoot names where to write the audit log (target-repo-local), but
  // patterns always load from the conduit install itself.
  const logRoot = repoRoot ?? findRepoRoot();
  const patternsRoot = getConduitInstallRoot();

  let patternsFile: PatternsFile;
  try {
    const raw = fs.readFileSync(path.join(patternsRoot, 'security', 'sanitizer', 'patterns.yaml'), 'utf-8');
    patternsFile = yaml.load(raw) as PatternsFile;
  } catch {
    // Fail closed if patterns unavailable
    appendLog(logRoot, commandName, input, 'block', ['sanitizer_unavailable']);
    throw new Error('sanitizer unavailable: fail closed — patterns.yaml missing from conduit install');
  }

  const matches: string[] = [];
  let decision = 'allow';

  for (const [groupName, group] of Object.entries(patternsFile.patterns)) {
    // Context flooding check
    if (group.max_input_chars !== undefined && input.length > group.max_input_chars) {
      matches.push(`${groupName}:context_flooding`);
      const action = patternsFile.actions[group.severity] ?? 'block_and_escalate';
      decision = resolveDecision(decision, action);
      continue;
    }

    if (!group.patterns) continue;

    for (const pattern of group.patterns) {
      try {
        const re = new RegExp(pattern, 'i');
        if (re.test(input)) {
          matches.push(`${groupName}:${pattern}`);
          const action = patternsFile.actions[group.severity] ?? 'block_and_escalate';
          decision = resolveDecision(decision, action);
        }
      } catch {
        // Invalid regex in patterns file — skip and log
        matches.push(`${groupName}:invalid_pattern`);
      }
    }
  }

  const allowed = decision === 'allow';
  appendLog(logRoot, commandName, input, decision, matches);

  return {
    allowed,
    sanitized: input,
    decision,
    matches,
  };
}

// block_and_escalate wins over sanitize_and_log wins over allow
function resolveDecision(current: string, incoming: string): string {
  const priority: Record<string, number> = { block_and_escalate: 2, sanitize_and_log: 1, allow: 0 };
  const currentP = priority[current] ?? 0;
  const incomingP = priority[incoming] ?? 0;
  return incomingP > currentP ? incoming : current;
}

function appendLog(repoRoot: string, commandName: string, input: string, decision: string, patterns: string[]): void {
  try {
    const hash = crypto.createHash('sha256').update(input).digest('hex');
    const entry = `${new Date().toISOString()} command=${commandName} input_sha256=${hash} decision=${decision} patterns=${patterns.join(',')}`;
    const logPath = path.join(repoRoot, '.conduit', 'sanitizer.log');
    fs.mkdirSync(path.dirname(logPath), { recursive: true });
    fs.appendFileSync(logPath, entry + '\n', 'utf-8');
  } catch {
    // Log failure is non-fatal — do not surface to caller
  }
}

export function findRepoRoot(startDir?: string): string {
  let current = startDir ?? process.cwd();
  while (true) {
    if (fs.existsSync(path.join(current, '.conduit'))) return current;
    const parent = path.dirname(current);
    if (parent === current) throw new Error('repo root not found (.conduit directory not found in any parent)');
    current = parent;
  }
}
